src/codec/mod.rs

```diff
@@ -0,0 +1,158 @@
+//! A collection of codecs that can be used to transform between bytes streams /
+//! byte messages, byte frames and structured events.
+
+#![deny(missing_docs)]
+
+mod framers;
+mod parsers;
+
+use crate::{
+    event::Event,
+    internal_events::{DecoderFramingFailed, DecoderParseFailed},
+};
+use bytes::BytesMut;
+pub use framers::*;
+pub use parsers::*;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+
+/// An error that occurred while decoding structured events from a byte stream /
+/// byte messages.
+#[derive(Debug)]
+pub enum Error {
+    /// The error occurred while producing byte frames from the byte stream /
+    /// byte messages.
+    FramingError(BoxedFramingError),
+    /// The error occurred while parsing structured events from a byte frame.
+    ParsingError(crate::Error),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::FramingError(error) => write!(formatter, "FramingError({})", error),
+            Self::ParsingError(error) => write!(formatter, "ParsingError({})", error),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Self::FramingError(Box::new(error))
+    }
+}
+
+#[derive(Clone)]
+/// A decoder that can decode structured events from a byte stream / byte
+/// messages.
+pub struct Decoder {
+    framer: BoxedFramer,
+    parser: BoxedParser,
+}
+
+impl Decoder {
+    /// Creates a new `Decoder` with the specified `Framer` to produce byte
+    /// frames from the byte stream / byte messages and `Parser` to parse
+    /// structured events from a byte frame.
+    pub fn new(framer: BoxedFramer, parser: BoxedParser) -> Self {
+        Self { framer, parser }
+    }
+}
+
+impl tokio_util::codec::Decoder for Decoder {
+    type Item = (SmallVec<[Event; 1]>, usize);
+    type Error = Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        // Frame bytes from the incoming byte stream / byte messages.
+        let frame = self.framer.decode(buf).map_err(|error| {
+            emit!(DecoderFramingFailed { error: &error });
+            Error::FramingError(error)
+        })?;
+
+        let frame = match frame {
+            Some(frame) => frame,
+            _ => return Ok(None),
+        };
+
+        let byte_size = frame.len();
+
+        // Parse structured events from the byte frame.
+        self.parser
+            .parse(frame)
+            .map(|event| Some((event, byte_size)))
+            .map_err(|error| {
+                emit!(DecoderParseFailed { error: &error });
+                Error::ParsingError(error)
+            })
```

Comment:
It seems like we should be able to deduplicate these method bodies without resorting to the higher-level function stuff you had before. What about something like:

```rust 
fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
    self.handle_framing_result(self.framer.decode(buf))
}

fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
    self.handle_framing_result(self.framer.decode_eof(buf))
}
```

And then in `impl Decoder`:

```rust
fn handle_framing_result(
    &mut self,
    result: Result<Option<Bytes>, BoxedFramingError>,
) -> Result<Option<Self::Item>, Self::Error> {
    // everything else
}
```
Still not super pretty, but simple enough and minimal duplication.

Reply:
Oh, yeah that's also a great to split them up without needing a closure there!

Had to use an intermediate result for `frame` since `self` can't be mutably borrowed twice